[SPARK-23366] Improve hot reading path in ReadAheadInputStream

juliuszsompolski opened the pull request:
## What changes were proposed in this pull request?

`ReadAheadInputStream` was introduced in https://github.com/apache/spark/pull/18317/ to optimize reading spill files from disk.
However, from the profiles it seems that the hot path of reading small amounts of data (like readInt) is inefficient - it involves taking locks, and multiple checks.

Optimize locking: Lock is not needed when simply accessing the active buffer. Only lock when needing to swap buffers or trigger async reading, or get information about the async state.

Optimize short-path single byte reads, that are used e.g. by Java library DataInputStream.readInt.

The asyncReader used to call "read" only once on the underlying stream, that never filled the underlying buffer when it was wrapping an LZ4BlockInputStream. If the buffer was returned unfilled, that would trigger the async reader to be triggered to fill the read ahead buffer on each call, because the reader would see that the active buffer is below the refill threshold all the time.

However, filling the full buffer all the time could introduce increased latency, so also add an `AtomicBoolean` flag for the async reader to return earlier if there is a reader waiting for data.

Remove `readAheadThresholdInBytes` and instead immediately trigger async read when switching the buffers. It allows to simplify code paths, especially the hot one that then only has to check if there is available data in the active buffer, without worrying if it needs to retrigger async read. It seems to have positive effect on perf.

## How was this patch tested?

It was noticed as a regression in some workloads after upgrading to Spark 2.3. 

It was particularly visible on TPCDS Q95 running on instances with fast disk (i3 AWS instances).
Running with profiling:
* Spark 2.2 - 5.2-5.3 minutes 9.5% in LZ4BlockInputStream.read
* Spark 2.3 - 6.4-6.6 minutes 31.1% in ReadAheadInputStream.read
* Spark 2.3 + fix - 5.3-5.4 minutes 13.3% in ReadAheadInputStream.read - very slightly slower, practically within noise.

We didn't see other regressions, and many workloads in general seem to be faster with Spark 2.3 (not investigated if thanks to async readed, or unrelated).

juliuszsompolski committed:
locking tweak

juliuszsompolski committed:
fill the read ahead buffer

juliuszsompolski commented:
cc @kiszk @sitalkedia @zsxwing

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution/736/
Test PASSed.

SparkQA commented:
**[Test build #87243 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/87243/testReport)** for PR 20555 at commit [`b26ffce`](https://github.com/apache/spark/commit/b26ffce6780078dbc38bff658e1ef7e9c56c3dd8).

SparkQA commented:
**[Test build #87243 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/87243/testReport)** for PR 20555 at commit [`b26ffce`](https://github.com/apache/spark/commit/b26ffce6780078dbc38bff658e1ef7e9c56c3dd8).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/87243/
Test PASSed.

kiszk commented on core/src/main/java/org/apache/spark/io/ReadAheadInputStream.java:
```diff
@@ -232,7 +229,9 @@ private void waitForAsyncReadComplete() throws IOException {
       while (readInProgress) {
+        isWaiting.set(true);
         asyncReadComplete.await();
+        isWaiting.set(false);
```
What happens if `await()` throws an exception? Is it ok not to update `isWaiting`?

juliuszsompolski replied:
Good catch, I added `isWaiting.set(false)` to the finally branch.
Actually, since the whole implementation assumes that there is only one reader, I removed the while() loop, since there is no other reader to race with us to trigger another read.

In practice I think not updating `isWaiting` it would have been benign, as after the exception the query will be going down with an `InterruptedException`, or elsewise anyone upstream handling that exception would most probably declare that stream as unusable afterwards anyway.

kiszk commented on core/src/main/java/org/apache/spark/io/ReadAheadInputStream.java:
```diff
@@ -258,54 +262,43 @@ public int read(byte[] b, int offset, int len) throws IOException {
+            return -1;
+          }
         }
+        // Swap the newly read read ahead buffer in place of empty active buffer.
```
Is it good to use `read-ahead` in comments for ease of reading?

juliuszsompolski replied:
Other existing places in comments in the file use `read ahead`.

juliuszsompolski committed:
reset isWaiting after exception

SparkQA commented:
**[Test build #87337 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/87337/testReport)** for PR 20555 at commit [`ca45a88`](https://github.com/apache/spark/commit/ca45a8887ad05e1a4ec4e690ebd7dfa2ca3a3968).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution/812/
Test PASSed.

juliuszsompolski committed:
remove waiting loop

AmplabJenkins commented:
Merged build finished. Test PASSed.

juliuszsompolski committed:
add short path for skip

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution/813/
Test PASSed.

SparkQA commented:
**[Test build #87339 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/87339/testReport)** for PR 20555 at commit [`eaa6b4e`](https://github.com/apache/spark/commit/eaa6b4e8de8415f01724d0341825f7aa9a63f81e).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution/814/
Test PASSed.

SparkQA commented:
**[Test build #87340 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/87340/testReport)** for PR 20555 at commit [`7238181`](https://github.com/apache/spark/commit/723818102cb84556598e69eaf12b255bcf9ada6a).

SparkQA commented:
**[Test build #87339 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/87339/testReport)** for PR 20555 at commit [`eaa6b4e`](https://github.com/apache/spark/commit/eaa6b4e8de8415f01724d0341825f7aa9a63f81e).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/87339/
Test FAILed.

SparkQA commented:
**[Test build #87340 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/87340/testReport)** for PR 20555 at commit [`7238181`](https://github.com/apache/spark/commit/723818102cb84556598e69eaf12b255bcf9ada6a).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/87340/
Test FAILed.

juliuszsompolski committed:
fix compilation

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution/818/
Test PASSed.

SparkQA commented:
**[Test build #87346 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/87346/testReport)** for PR 20555 at commit [`d6d44fc`](https://github.com/apache/spark/commit/d6d44fc3007205437a58e2eb9333bb055598dba1).

gatorsmile commented:
Also cc @jiangxb1987 @cloud-fan

SparkQA commented:
**[Test build #87337 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/87337/testReport)** for PR 20555 at commit [`ca45a88`](https://github.com/apache/spark/commit/ca45a8887ad05e1a4ec4e690ebd7dfa2ca3a3968).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/87337/
Test PASSed.

SparkQA commented:
**[Test build #87346 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/87346/testReport)** for PR 20555 at commit [`d6d44fc`](https://github.com/apache/spark/commit/d6d44fc3007205437a58e2eb9333bb055598dba1).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/87346/
Test PASSed.

cloud-fan commented:
LGTM

cloud-fan commented on core/src/main/java/org/apache/spark/io/ReadAheadInputStream.java:
```diff
@@ -258,54 +263,43 @@ public int read(byte[] b, int offset, int len) throws IOException {
-        if (isEndOfStream()) {
-          return -1;
+        if (!readAheadBuffer.hasRemaining()) {
+          // The first read or activeBuffer is skipped.
```
unrelated question: what does `activeBuffer is skipped` mean?

juliuszsompolski replied:
skipped using `skip()`.
I moved the comment over from a few lines above, but looking at `skip()` now I don't think it can happen - the skip would trigger an `readAsync` read in that case.
I'll update the comment.

cloud-fan commented:
cc @sameeragarwal @sitalkedia , shall we have this in Spark 2.3?

jiangxb1987 reviewed (approved):
LGTM

jiangxb1987 commented:
It would also be great if we can add some unit tests on the read ahead stream model.

sitalkedia reviewed (approved):
LGTM. Thanks for fixing this @juliuszsompolski .

juliuszsompolski committed:
update comment

SparkQA commented:
**[Test build #87415 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/87415/testReport)** for PR 20555 at commit [`5273176`](https://github.com/apache/spark/commit/5273176abf20c8a87776df21fa7458b33f11a990).

juliuszsompolski committed:
update test suite with uneven buffer sizes

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution/868/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution/869/
Test PASSed.

juliuszsompolski committed:
more testing combinations

juliuszsompolski commented:
@jiangxb1987 there is ReadAheadInputStreamSuite that extends GenericFileInputStreamSuite.
I updated it and added more combination testing with different buffer sizes that should exercise more interactions between the wrapped and outer buffers.

SparkQA commented:
**[Test build #87422 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/87422/testReport)** for PR 20555 at commit [`1b3e970`](https://github.com/apache/spark/commit/1b3e9709db730b2cb5caa155ecb4562c18044f73).

SparkQA commented:
**[Test build #87415 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/87415/testReport)** for PR 20555 at commit [`5273176`](https://github.com/apache/spark/commit/5273176abf20c8a87776df21fa7458b33f11a990).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

SparkQA commented:
**[Test build #87416 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/87416/testReport)** for PR 20555 at commit [`62cefcd`](https://github.com/apache/spark/commit/62cefcd6e97e5826713b6690f62b542a82d9e2aa).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/87416/
Test PASSed.

zsxwing commented on core/src/main/java/org/apache/spark/io/ReadAheadInputStream.java:
```diff
@@ -78,9 +79,8 @@
-  // we issue an async read from the underlying input stream.
-  private final int readAheadThresholdInBytes;
+  // whether there is a reader waiting for data.
+  private AtomicBoolean isWaiting = new AtomicBoolean(false);
```
You can just use `volatile` here

juliuszsompolski replied:
I'll leave it be - should compile to basically the same, and with using `AtomicBoolean` the intent seems more readable to me.

zsxwing commented:
Looks pretty good. Left two minor comments.

SparkQA commented:
**[Test build #87422 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/87422/testReport)** for PR 20555 at commit [`1b3e970`](https://github.com/apache/spark/commit/1b3e9709db730b2cb5caa155ecb4562c18044f73).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/87422/
Test PASSed.

juliuszsompolski committed:
while loop against spurious wakeups

juliuszsompolski commented on core/src/main/java/org/apache/spark/io/ReadAheadInputStream.java:
```diff
@@ -230,24 +227,32 @@ private void signalAsyncReadComplete() {
+    isWaiting.set(true);
     try {
-      while (readInProgress) {
+      if (readInProgress) {
```
Good catch, thanks!

SparkQA commented:
**[Test build #87435 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/87435/testReport)** for PR 20555 at commit [`52f4a7c`](https://github.com/apache/spark/commit/52f4a7c3e97c475b4464b82c7d8e00dcd9d889b3).

zsxwing commented:
LGTM

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution/883/
Test PASSed.

cloud-fan commented on core/src/main/java/org/apache/spark/io/ReadAheadInputStream.java:
```diff
@@ -230,6 +227,7 @@ private void signalAsyncReadComplete() {
     stateChangeLock.lock();
+    isWaiting.set(true);
     try {
       while (readInProgress) {
```
shall we add a comment about spurious wakeup? Otherwise someone else may still mistakenly remove it in the future.

juliuszsompolski committed:
add comment about spuriour wakeups

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution/886/
Test PASSed.

SparkQA commented:
**[Test build #87438 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/87438/testReport)** for PR 20555 at commit [`b6852aa`](https://github.com/apache/spark/commit/b6852aa8a7f0e053985d5b89ba4020792d648f82).

SparkQA commented:
**[Test build #87438 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/87438/testReport)** for PR 20555 at commit [`b6852aa`](https://github.com/apache/spark/commit/b6852aa8a7f0e053985d5b89ba4020792d648f82).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/87438/
Test PASSed.

jiangxb1987 commented:
LGTM

cloud-fan commented:
thanks, merging to master!
